rust/src/dataset/cleanup.rs

```diff
@@ -0,0 +1,834 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A task to clean up a lance dataset, removing files that are no longer
+//! needed.
+//!
+//! Currently we try and be rather conservative about what we delete.
+//!
+//! The following types of files may be deleted by the cleanup function:
+//!
+//! * Old manifest files - If a manifest file is older than the threshold
+//!   and is not the latest manifest then it will be deleted.
+//! * Unreferenced data files - If a data file is not referenced by any
+//!   fragment in a valid manifest file then it will be deleted.
+//! * Unreferenced delete files - If a delete file is not referenced by
+//!   any fragment in a valid manifest file then it will be deleted.
+//! * Unreferenced index files - If an index file is not referenced by
+//!   any valid manifest file then it will be deleted.
+//!
+//! It is also impossible to distinguish between a data file which was
+//! leftover from an abandoned transaction and a data file which is part
+//! of an ongoing write operation (both will be unreferenced data files).
+//! To solve this problem we only delete data files that have been unmodified
+//! since the delete threshold.  The delete threshold must be at least 7
+//! days and we assume that any ongoing writes will complete within 7 days.
+
+use std::{
+    collections::HashSet,
+    future,
+    sync::{Arc, Mutex},
+};
+
+use chrono::{DateTime, Duration, Utc};
+use futures::{stream, StreamExt, TryStreamExt};
+use object_store::path::Path;
+
+use crate::{
+    format::Manifest,
+    io::{deletion_file_path, read_manifest, reader::read_manifest_indexes},
+    utils::temporal::utc_now,
+    Dataset, Error, Result,
+};
+
+#[derive(Clone, Debug, Default)]
+struct ReferencedFiles {
+    data_paths: HashSet<Path>,
+    delete_paths: HashSet<Path>,
+    tx_paths: HashSet<Path>,
+    index_uuids: HashSet<String>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct RemovalStats {
+    unreferenced_data_paths: u64,
```

Comment:
We'd have to change the read_dir_all method to return sizes along with the paths.  Not hard, but then it would be out of sync with the singular read.  I'll defer for a follow-up in case it results in bike shedding on API :)

Reply:
https://github.com/lancedb/lance/issues/1175